File: src/services/image-optimization.service.ts
import { Injectable } from '@angular/core';
import { getStandardWikimediaSize } from '../constants/image-sizes';

/**
 * Service to handle image URL standardization and optimization.
 */
@Injectable({
  providedIn: 'root'
})
export class ImageOptimizationService {

  /**
   * Standardizes a Wikipedia thumbnail URL to use recommended production sizes.
   * Example: https://upload.wikimedia.org/wikipedia/commons/thumb/1/1a/File.jpg/640px-File.jpg
   * becomes: https://upload.wikimedia.org/wikipedia/commons/thumb/1/1a/File.jpg/960px-File.jpg
   * 
   * @param url The Wikipedia thumbnail URL to standardize.
   * @param targetWidth Optional target width. If not provided, it snaps current size to the next standard bucket.
   * @returns The standardized URL.
   */
  standardizeWikipediaUrl(url: string, targetWidth?: number): string {
    if (!url.includes('upload.wikimedia.org')) {
      return url;
    }

    // Match the size part of the URL, e.g., /640px-
    const sizeMatch = url.match(/\/(\d+)px-/);
    if (!sizeMatch) {
      // Not a standard thumbnail URL or already at original resolution
      return url;
    }

    const currentWidth = parseInt(sizeMatch[1], 10);
    const widthToUse = targetWidth !== undefined ? targetWidth : currentWidth;
    const standardWidth = getStandardWikimediaSize(widthToUse);

    // Replace the size in the URL
    return url.replace(`/${currentWidth}px-`, `/${standardWidth}px-`);
  }
}

File: pocketgull_flutter/lib/services/image_optimization_service.dart
/// Image Optimization Service — Wikipedia thumbnail URL standardization.
///
/// Flutter parity with Angular `image-optimization.service.ts` (55 lines).
/// Snaps Wikimedia thumbnail widths to standard production buckets.
library;

/// Standard Wikimedia thumbnail size buckets.
const _standardSizes = [320, 480, 640, 800, 960, 1280, 1920];

/// Returns the next standard Wikimedia thumbnail size >= [width].
int getStandardWikimediaSize(int width) {
  for (final size in _standardSizes) {
    if (size >= width) return size;
  }
  return _standardSizes.last;
}

class ImageOptimizationService {
  /// Standardizes a Wikipedia thumbnail URL to use recommended production sizes.
  String standardizeWikipediaUrl(String url, {int? targetWidth}) {
    try {
      final uri = Uri.parse(url);
      if (uri.host != 'upload.wikimedia.org') return url;
    } catch (_) {
      return url;
    }

    // Match the size part of the URL, e.g., /640px-
    final sizeMatch = RegExp(r'/(\d+)px-').firstMatch(url);
    if (sizeMatch != null) {
      final currentWidth = int.parse(sizeMatch.group(1)!);
      final widthToUse = targetWidth ?? currentWidth;
      final standardWidth = getStandardWikimediaSize(widthToUse);
      return url.replaceFirst('/${currentWidth}px-', '/${standardWidth}px-');
    }

    // If it's an original image (doesn't contain '/thumb/'), rewrite it.
    if (!url.contains('/thumb/')) {
      final match = RegExp(
        r'(https://upload\.wikimedia\.org/wikipedia/[^/]+)/(.+)/([^/]+)$',
      ).firstMatch(url);
      if (match != null) {
        final widthToUse = targetWidth ?? 640;
        final standardWidth = getStandardWikimediaSize(widthToUse);
        return '${match.group(1)}/thumb/${match.group(2)}/${match.group(3)}/${standardWidth}px-${match.group(3)}';
      }
    }

    return url;
  }
}
